src/net.rs

```diff
@@ -187,13 +191,30 @@ impl Builder {
 
     /// Spawn a gossip actor and get a handle for it
     pub async fn spawn(self, endpoint: Endpoint) -> Result<Gossip, Error> {
-        let addr = endpoint.node_addr().await?;
+        // We want to wait for our endpoint to be addressable by other nodes before launching gossip,
+        // because otherwise our Join messages, which will be forwarded into the swarm through a random
+        // walk, might not include an address to talk back to us.
+        // `Endpoint::node_addr` always waits for direct addresses to be available, which never completes
+        // when running as WASM in browser. Therefore, we instead race the futures that wait for the direct
+        // addresses or the home relay to be initialized, and construct our node address from that.
+        // TODO: Make `Endpoint` provide a more straightforward API for that.
+        let addr = {
+            n0_future::future::race(
+                endpoint.direct_addresses().initialized().map(|_| ()),
+                endpoint.home_relay().initialized().map(|_| ()),
+            )
+            .await;
+            let addrs = endpoint.direct_addresses().get()?.unwrap_or_default();
+            let addrs = addrs.into_iter().map(|x| x.addr);
+            let home_relay = endpoint.home_relay().get()?;
```

Comment:
I'd say it's fine to use `.unwrap()` on these here, since we know we can't get a `Disconnected` as long as we keep an `Endpoint` alive at the same time.
Alternatively we can make it `.expect("endpoint alive")` to explain better?
Anyhow. Not that important as long as we're using `anyhow` still.

Reply:
changed :+1: